qiskit/transpiler/passes/basis/basis_translator.py

```diff
@@ -210,100 +197,152 @@ def _basis_search(equiv_lib, source_basis, target_basis, heuristic):
+    for key in equiv_lib._get_all_keys():
+        target = get_nid_or_insert(key)
+        all_gates_in_lib.add(key)
+        for equiv in equiv_lib._get_equivalences(key):
+            sources = set(
+                Key(name=gate.name, num_qubits=len(qargs)) for gate, qargs, _ in equiv.circuit
+            )
+            all_gates_in_lib |= sources
+            edges = [
                 (
-                    frozenset(
-                        basis_remain
-                        | {(inst.name, inst.num_qubits) for inst, qargs, cargs in equiv.data}
-                    ),
-                    params,
-                    equiv,
+                    get_nid_or_insert(source),
+                    target,
+                    {"index": rcounter, "rule": equiv, "source": source},
                 )
-                for params, equiv in equivs
+                for source in sources
             ]
 
-            # Weight total path length of transformation weakly.
-            tentative_cost_from_source = cost_from_source[current_basis] + 1e-3
-
-            for neighbor, params, equiv in neighbors:
-                if neighbor in closed_set:
-                    continue
-
-                if tentative_cost_from_source >= cost_from_source[neighbor]:
-                    continue
-
-                open_set.add(neighbor)
-                came_from[neighbor] = (current_basis, gate_name, gate_num_qubits, params, equiv)
-                cost_from_source[neighbor] = tentative_cost_from_source
-                est_total_cost[neighbor] = tentative_cost_from_source + heuristic(
-                    neighbor, target_basis
-                )
-                heappush(open_heap, (est_total_cost[neighbor], next(basis_count), neighbor))
-
-    return None
+            num_gates_for_rule[rcounter] = len(sources)
+            graph.add_edges_from(edges)
+            rcounter += 1
+
+    # This is only neccessary since gates in target basis are currently reported by
+    # their names and we need to have in addition the number of qubits they act on.
+    target_basis_keys = [
+        key
+        for gate in target_basis
+        for key in filter(lambda key, name=gate: key.name == name, all_gates_in_lib)
+    ]
+
+    vis = BasisSearchVisitor(graph, source_basis, target_basis_keys, num_gates_for_rule)
+    # we add a dummy node and connect it with gates in the target basis.
+    # we'll start the search from this dummy node.
+    dummy = graph.add_node("dummy starting node")
+    graph.add_edges_from_no_data([(dummy, nodes_to_indices[key]) for key in target_basis_keys])
+    rtn = None
+    try:
+        retworkx.digraph_dijkstra_search(graph, [dummy], vis.edge_cost, vis)
+    except StopIfBasisRewritable:
+        rtn = vis.basis_transforms
+
+        logger.debug("Transformation path:")
+        for gate_name, gate_num_qubits, params, equiv in rtn:
+            logger.debug("%s/%s => %s\n%s", gate_name, gate_num_qubits, params, equiv)
+
```

Comment:
Since we're logging, should we have `try/else` block to log a debug message on a failed translation?

Reply:
Well, if translation fails we'll raise a `TranspilerError` immediately and there isn't something useful to log at this point. That being said, I still need to improve log messages during the search.